src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,431 @@
+            if (!HasResponseStarted)
+            {
+                return WriteAsyncAwaited(data, cancellationToken);
+            }
+
+            lock (_stateSync)
+            {
+                DisableReads();
+                return Output.WriteAsync(data, cancellationToken: cancellationToken);
+            }
+        }
+
+        /// <summary>
+        /// Flushes the data in the output pipe
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await InitializeResponse(0);
+            await Output.FlushAsync(cancellationToken);
+        }
+
+        public void StartProcessingRequestAndResponseBody()
+        {
+            if (_processBodiesTask == null)
+            {
+                lock (_createReadWriteBodySync)
+                {
+                    if (_processBodiesTask == null)
+                    {
+                        _processBodiesTask = ConsumeAsync();
+                    }
+                }
+            }
+        }
+
+        private async Task WriteAsyncAwaited(ArraySegment<byte> data, CancellationToken cancellationToken)
+        {
+            await InitializeResponseAwaited(data.Count);
+
+            // WriteAsyncAwaited is only called for the first write to the body.
+            // Ensure headers are flushed if Write(Chunked)Async isn't called.
+
+            Task writeTask;
+            lock (_stateSync)
+            {
+                DisableReads();
+
+                // Want to guarantee that data has been written to the pipe before releasing the lock.
+                writeTask = Output.WriteAsync(data, cancellationToken: cancellationToken);
+            }
+            await writeTask;
+        }
+
+        // ConsumeAsync is called when either the first read or first write is done. 
+        // There are two modes for reading and writing to the request/response bodies.
+        // 1. Await all reads and try to read from the Output pipe
+        // 2. Done reading and await all writes.
+        private async Task ConsumeAsync()
+        {
+            await ReadAndWriteLoopAsync();
+
+            // If we are done writing, complete the output pipe and return
+            // Input Pipe will be closed when ReadAndWriteLoopAsync returns
+            if (_doneWriting)
+            {
+                Output.Reader.Complete();
+                return;
+            }
+
+            await WriteLoopAsync();
+        }
+
+        private unsafe IISAwaitable ReadFromIISAsync(int length)
+        {
+            lock (_stateSync)
+            {
+                // We don't want to read if there is data available in the output pipe
+                if (Output.Reader.TryRead(out var result) && !result.Buffer.IsEmpty)
+                {
+                    _operation.Complete(hr: IISServerConstants.HResultCancelIO, cbBytes: 0);
```

Comment:
I would add a comment here describing that we're marking the current  operation as cancelled as a result.